tests/compile-fail/diverging_sub_expression.rs

```diff
@@ -0,0 +1,14 @@
+#![feature(plugin, never_type)]
+#![plugin(clippy)]
+#![deny(diverging_sub_expression)]
+
+#[allow(empty_loop)]
+fn diverge() -> ! { loop {} }
+
+#[allow(unused_variables, unnecessary_operation)]
+fn main() {
+    let b = true;
+    b || diverge(); //~ ERROR sub-expression diverges
+    let y = (5, diverge(), 6); //~ ERROR sub-expression diverges
+    println!("{}", y.1); //~ ERROR sub-expression diverges
```

Comment:
I guess this is because `y: (_, !, _)`, but that last error is rather confusing. I don't think you can get there without diverging sooner while constructing `y` which must involve a diverging sub-expression.